### ISSUE TITLE
Do not set HttpContext.Current!! This can cause so many issues

### ISSUE THREAD

**Issue by Shazwazza** [leekelleher/umbraco-mortar#13]
Currently you have this:

https://github.com/leekelleher/umbraco-mortar/blob/master/Src/Our.Umbraco.Mortar.Courier/DataResolvers/MortarDataResolver.cs#L42

Which is entirely not necessary for ensuring an UmbracoContext. Setting the HttpContext.Current can cause so many issues:

* You don't know what is relying on HttpContext.Current - if things are running on a background thread and do specific checks for this, behaviors will change
* This is the specific reason for this courier issue: http://issues.umbraco.org/issue/UMB-187 
     * The NH sessions are controlled by whether it's a web request or not, initial extraction runs on a background thread (no web request) but because your data resolver is created, suddenly HttpContext.Current is not null and therefore the NH transaction is no longer valid right in the middle of processing
* Unless you are working between threads/background threads and know specifically that you are transfering an existing HttpContext to another thread (which is still not good practice in general), you should never set HttpContext.Current since there will be repercussions.
* This would also cause issues for whenever there are request based lifetime objects, since they are stored in HttpContext, if suddenly one was available, any request based lifetime objects would check or store items there, even though originally they assumed they were application based (i.e. not possible to run web based).

**Comment by leekelleher** [comment]
Thanks for raising this @Shazwazza!

Yeah, I knew setting the `HttpContext.Current` was a nasty hack - at the time I was stuck on how to deal with needing the `UmbracoContext`. (I added comments above the code)

re: [#UMB-187](http://issues.umbraco.org/issue/UMB-187) - I don't have permission to view the ticket - but doesn't matter - I get the idea about NH threads/context.

So the question is... how do we deal with ensuring the `UmbracoContext`?

**Comment by Shazwazza** [comment]
You just create your own http ctx instance and pass it in, you don't need
to use the http context Singleton and never set that Singleton :)

**Comment by leekelleher** [comment]
Gotcha! (_of course, makes sense now!_)

I'll update our code and test it out.  Thanks again.

**Comment by sitereactor** [comment]
I'll just add to this that the Mortar Courier provider causes problems while Courier runs its extraction process.
Clone a fresh *insert brand name here* site, start it in IIS and you'll see exceptions in the CourierTraceLog.txt. Now, do the same but remove the the `Our.Umbraco.Mortar.dll` from the bin folder before starting the site and you'll see that everything works - no exceptions are thrown during extraction.

It's difficult to say what other problems this is causing during extraction when deploying between environments, but I'm certain it has an impact.

**Comment by leekelleher** [comment]
OK, all noted.  _Obviously, I was unaware of the wider impact of this._

Here is my proposed code change... does that work better?

```csharp
if (UmbracoContext.Current == null)
{
	var httpContext = HttpContext.Current != null
		? HttpContext.Current
		: new HttpContext(new HttpRequest(string.Empty, "http://tempuri.org", string.Empty), new HttpResponse(new StringWriter()));

	UmbracoContext.EnsureContext(new HttpContextWrapper(httpContext), ApplicationContext.Current, true);
}
```

**Comment by Shazwazza** [comment]
Yup, that works :)

The next question is where are you using this UmbracoContext? and since you are passing in 'true' it will assign the UmbracoContext.Current singleton

**Comment by leekelleher** [comment]
OK, here's the walk-through...

I'm deserializing the JSON, via Mortar's property-value converter:
https://github.com/leekelleher/umbraco-mortar/blob/master/Src/Our.Umbraco.Mortar.Courier/DataResolvers/MortarDataResolver.cs#L105

Which then checks which type of block it is using (richtext, doctype, etc), then tries to construct the "fake" DocType:
https://github.com/leekelleher/umbraco-mortar/blob/master/Src/Our.Umbraco.Mortar/ValueConverters/MortarValueConverter.cs#L92

Which then makes a call to `PublishedProperty.GetDetached` (uses reflection to call core internal):
https://github.com/leekelleher/umbraco-mortar/blob/master/Src/Our.Umbraco.Mortar/ValueConverters/MortarValueConverter.cs#L170

To which `GetDetached` requires `UmbracoContext`:
https://github.com/umbraco/Umbraco-CMS/blob/dev-v7/src/Umbraco.Web/Models/PublishedProperty.cs#L27

---

As for passing in `true` to replace the context... I haven't tried it as `false` yet... what would you recommend?

**Comment by Shazwazza** [comment]
Right ok, so it needs the singleton, so i guess you'll have to keep the 'true' parameter.

Been asking Stephen to make sure that all extension methods have overloads to pass in all of their dependencies - we need to stop relying on singletons, it makes things messy, un-testable and causes issues like this. Ideally, you'd just create a standalone UmbracoContext and just pass that instance to the methods you require. I know that's not always possible, but that's the ideal behavior.

**Comment by leekelleher** [comment]
Closing this ticket as resolved in PR #15
